### PR TITLE
[CLEANUP] Don't cache length property in for loops

### PR DESCRIPTION
**Pull request by topaxi**
Just like in https://github.com/emberjs/data/pull/4015

### CODE REVIEW

**Comment by topaxi** [comment]
I'm not sure if the failed tests are due to this change or some weird IE9 behavior. It seems to timeout in two tests which seem unrelated to these changes.

**Comment by rwjblue** [comment]
I will try to review in more detail later tonight/tomorrow, but one thing I wanted to make sure you took into account was that sometimes caching the length is actually needed.  Specifically, when you are in a loop and you are mutating that array inside the loop you may need to cache `length` (depending on what you are doing).  Did you review the surrounding code to confirm that is not the case in these instances?

File: packages/ember-metal/lib/chains.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ ChainWatchers.prototype = {
   remove(key, node) {
     let nodes = this.chains[key];
     if (nodes) {
-      for (var i = 0, l = nodes.length; i < l; i++) {
+      for (var i = 0; i < nodes.length; i++) {
         if (nodes[i] === node) {
           nodes.splice(i, 1);
```
Comment (topaxi) [inline packages/ember-metal/lib/chains.js]: Self note: This loop length should probably be cached.

**Comment by topaxi** [comment]
I did review surrounding code and it seemed that all those loops never mutated the iterated array / array-like object. Unless there are some liverecord-like arrays involved, there shouldn't be any issues.

I might missed one or two, currently double-checking the whole commit myself.

File: packages/ember-metal/lib/chains.js (hunk 1 above)
Comment (topaxi) [inline packages/ember-metal/lib/chains.js]: On a second look, this shouldn't need to cache length as we break the loop on the next line, might be confusing though.

**Comment by topaxi** [comment]
Resolved merge conflicts and rebased on master.

**Comment by stefanpenner** [comment]
Seems good, we will just need to double check each change.

**Comment by rwjblue** [comment]
Thank you for your time on this! I'm going to go ahead and close this for now as it seems to have gone stale. I'm happy to reopen if you would like to pick this back up again.

**Comment by topaxi** [comment]
Rebased and pushed to the branch https://github.com/topaxi/ember.js/tree/dont-cache-length
Github does not seem to update this PR though, is this because the PR is now closed?

I'm happy to reopen a new PR and update or rebase this as needed, just ping me if new merge conflicts happen.

**Comment by rwjblue** [comment]
Weird, it won't let me reopen the PR (I was able to reopen at least one that I closed yesterday).

**Pull request closed without merging**